mlflow/genai/utils/trace_utils.py

```diff
@@ -763,3 +774,252 @@ def batch_link_traces_to_run(
                 return
 
             _logger.warning(f"Failed to link batch of traces to run: {e}")
+
+
+class ExtractedToolFunction(BaseModel):
+    """Extracted tool function definition for structured output extraction."""
+
+    name: str = Field(description="The function/tool name")
+    description: str = Field(default="", description="What the tool does")
+    parameters: dict[str, Any] = Field(
+        default_factory=dict, description="The JSON schema for parameters"
+    )
+
+    model_config = {"extra": "forbid"}
+
+
+class ExtractedTool(BaseModel):
+    """Extracted tool definition for structured output extraction."""
+
+    type: str = Field(default="function", description="Always 'function'")
+    function: ExtractedToolFunction = Field(description="The function definition")
+
+    model_config = {"extra": "forbid"}
+
+
+class ExtractedToolsFromTrace(BaseModel):
+    """Schema for extracting available tools from a trace using structured output."""
+
+    tools: list[ExtractedTool] = Field(
+        default_factory=list,
+        description="List of all available tools found in the trace",
+    )
+
+    model_config = {"extra": "forbid"}
+
+
+def extract_available_tools_from_trace(trace: Trace, model: str | None = None) -> list["ChatTool"]:
```

Comment:
```suggestion
def extract_available_tools_from_trace(trace: Trace | None, model: str | None = None) -> list["ChatTool"]:
```
nit: should this be `Trace | None` if we have to check for None? also, I think it is safe to assume if the trace exists, trace.data won't be none

Reply:
Removed the check for None